core/src/cluster_info.rs

```diff
@@ -1788,6 +1812,44 @@ impl ClusterInfo {
         }
     }
 
+    // Returns a predicate checking if the pull request is from a valid
+    // address, and if the address have responded to a ping request. Also
+    // appends ping packets for the addresses which need to be (re)verified.
+    fn check_pull_request<'a, R>(
+        &'a self,
+        now: Instant,
+        mut rng: &'a mut R,
+        packets: &'a mut Packets,
+    ) -> impl FnMut(&PullData) -> bool + 'a
+    where
+        R: Rng + CryptoRng,
+    {
+        let mut cache = HashMap::<(Pubkey, SocketAddr), bool>::new();
+        let mut pingf = move || Ping::new_rand(&mut rng, &self.keypair).ok();
+        let mut ping_cache = self.ping_cache.write().unwrap();
+        let mut hard_check = move |node| {
+            let (_check, ping) = ping_cache.check(now, node, &mut pingf);
+            if let Some(ping) = ping {
+                let ping = Protocol::PingMessage(ping);
+                let ping = Packet::from_data(&node.1, ping);
+                packets.packets.push(ping);
+            }
+            // TODO: For backward compatibility, this unconditionally returns
+            // true for now. It has to return _check, once nodes start
+            // responding to ping messages.
+            true
```

Comment:
We can plumb in the final code here as well, making it conditional on a new feature added to `runtime/src/feature_set.rs`.   This way we can dynamically activate this feature on clusters once 95% of the validator fleet has upgraded.   Right now `FeatureSet` is not plumbed though cluster_info though, so I suggest adding this plumbing first as a forerunner PR to this one.  `core/` should probably use the `FeatureSet` from the highest rooted bank.

Reply:
sure, working on it.

Reply:
updated the diff to use the feature-set added in https://github.com/solana-labs/solana/pull/12878